src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -360,6 +378,50 @@ private void AddAssemblies (ZipArchiveEx apk)
 					count = 0;
 				}
 			}
+
+			void EnsureCompressedAssemblyData (string sourcePath, uint descriptorIndex)
+			{
+				if (compressedAssembly == null)
+					compressedAssembly = new AssemblyCompression.AssemblyData (sourcePath, descriptorIndex);
+				else
+					compressedAssembly.SetData (sourcePath, descriptorIndex);
+			}
+
+			string CompressAssembly (ITaskItem assembly)
+			{
+				if (!compress) {
+					return assembly.ItemSpec;
+				}
+
+				if (bool.TryParse (assembly.GetMetadata ("AndroidSkipCompression"), out bool value) && value) {
+					Log.LogDebugMessage ($"Skipping compression of {assembly.ItemSpec} due to 'AndroidSkipCompression' == 'true' ");
+					return assembly.ItemSpec;
+				}
+
+				if (compressedAssembliesInfo.TryGetValue (Path.GetFileName (assembly.ItemSpec), out CompressedAssemblyInfo info) && info != null) {
+					EnsureCompressedAssemblyData (assembly.ItemSpec, info.DescriptorIndex);
+					AssemblyCompression.CompressionResult result = AssemblyCompression.Compress (compressedAssembly);
+					if (result != AssemblyCompression.CompressionResult.Success) {
+						switch (result) {
+							case AssemblyCompression.CompressionResult.EncodingFailed:
+								Log.LogWarning ($"Failed to compress {assembly.ItemSpec}");
```

Comment:
1. `Log.LogWarning()` calls should now *always* contain an associated `XA…` warning code.
2. Warning messages should be *actionable*.  This message is not actionable.

I thus believe that this should be `Log.LogMessage()` invocation.

Ditto the other `LogWarning()` invocations within this block.